### ISSUE TITLE
Rotated Bounding Boxes

### ISSUE THREAD

**Issue by edeyejedi** [ultralytics/yolov5#510]
## 🚀 Feature
As per title

## Motivation
Feature orientation is imperative to my application 

## Pitch
Train with rotated bounding boxes - will require some theta value in the Label file I guess
Detect the rotated boxes. 
Basically as discussed here for YV3
https://github.com/ultralytics/yolov3/issues/345

There is this:
https://github.com/ming71/rotate-yolov3
This is next level:
https://github.com/maudzung/Complex-YOLOv4-Pytorch
Then:
https://github.com/mrlooi/rotated_maskrcnn

I guess it would be really useful. 

## Alternatives
NA

## Additional context
Understand there isn't a magical coding fairy.......but man it would be cool, especially under YV5 :)

**Comment by glenn-jocher** [comment]
@edeyejedi that would be a cool addition! Anything labelable is trainable, but the constraint here is that you need to have rotation information in your labels, which is uncommon in most datasets. 

I'd say (my own cold calculus) rotation is among a few fringe applications that some people would like to see but really falls outside the mainstream enough that it's typically not worth complicating the codebase for the small userbase addition that would exploit the addons. Another one of these is hyperspectral image training.

Of course, if you have time and can implement (and verify) rotation training in a minimal manner I'll definitely have a serious look at the PR!

**Comment by glenn-jocher** [comment]
BTW @ming71 does some good work!

**Comment by edeyejedi** [comment]
thanks Glenn.
Yeah I have seen the version of @ming71, will have a crack at getting it up and running. 
Not YV5 though!  
Incorporation of rotation might be a bit beyond me at this stage!

**Comment by chmorgan** [comment]
I’m also interested in this feature. It seems particularly helpful when looking to use the bounding box to control a mechanical system.

**Comment by glenn-jocher** [comment]
@chmorgan do you have any particular dataset in mind? This is not really a difficult change, though many small changes would need to permeate throughout the repo (i.e. visualizations, IoU functions, NMS, etc), which togethor sum to a significant effort for a minority user base.

In any case we aim to make YOLOv5 as broadly applicable as feasibly possible.

**Comment by edeyejedi** [comment]
Thanks for sorting out my spelling @glenn-jocher 
@chmorgan I exchanged a few emails with @maudzung about his Complex-YOLOv4-Pytorch implementation, he seems confident of some changes he made to accommodate rotated bounded boxes. I have not had time to have a play with some commits he made, but it likely be a very good starting point to looking how it could be done in Yv5.

**Comment by glenn-jocher** [comment]
@edeyejedi sounds good. I myself don't have availability to work on this, but if you are interested in contributing, we are open to PRs!

**Comment by chmorgan** [comment]
@glenn-jocher the particular use case we are considering is using yolo to enable object trajectory tracking and interception by a servo mechanism. We would be tracking the object in the field of view and moving to grab it with an arm for an industrial use.

We don't have a particular public dataset available at the moment.

Is there a typical way this orientation issue is handled such as through another network that would be trained on the bounding box image + angle dataset only? I'm pretty much fresh out of the box when it comes to AI/ML so any direction would be much appreciated.

**Comment by chmorgan** [comment]
@edeyejedi interesting! We'd certainly be a user and possibly a contributor but lack the knowledge to lead such an effort at this time. Was hoping to draft someone that had more than a few days of yolo experience like I do :-)

**Comment by Ownmarc** [comment]
@chmorgan if I recall correctly, the heads of yolo can be modified to output one more value for the rotation, eg (-180 to 180 degree, scaled between something like 0 and 1). It will require experienced people and also a decent dataset as Glenn said, but its feasible :) 

There is also alot of tweaking to be done to get this to be general enough and relevant for other datasets.

**Comment by glenn-jocher** [comment]
@chmorgan yes, it's actually a simple change, xwyh regression output simply expands one element to accomodate a rotation value. You require data labelled with this extra rotation value obviously, since we are doing supervised learning here. If there are existing implementations the best pathway is probably to use their existing work or see if they would be willing to lend a hand here. 

Like I said before, it's not very difficult, its just that the small user base for this potential feature addition does not justify the effort on our part. We are trying to make the most impact with very limited resources, so we focus on the largest use cases, i.e. speed, accuracy, export, UI/UX.

**Comment by maudzung** [comment]
Hi,

Recently, I have adopted CenterNet ideas for the **rotated boxes** detection task. That's an anchor-free approach, no need for Non-max-suppression during inference. 

For the YOLO-based approach, we need to compute the **IoU of rotated boxes**. The IoU computation is **very expensive and slow** because we can't apply the vectorization. This leads to a slow training and inference process.
For the anchor-free method, it's faster training, faster inference. The number of parameters is much smaller than the YOLO models, so we can **avoid the overfitting problem** with a small dataset.

My implementation is [here](https://github.com/maudzung/Super-Fast-Accurate-3D-Object-Detection)

**Comment by edeyejedi** [comment]
hey @maudzung 
great work. I will have a look asap

**Comment by maudzung** [comment]
hi @edeyejedi 

One more advantage of the anchor-free method is that we don't need to find pre-defined anchors. Actually, It's difficult to find good anchors because we need to take the rotate angle into account.

**Comment by yushuinanrong** [comment]
@glenn-jocher 
This is the temporary workaround for me:
```
import skvideo.io
videogen = skvideo.io.vread(f.name)
```

**Comment by jmayank23** [comment]
> @chmorgan do you have any particular dataset in mind? This is not really a difficult change, though many small changes would need to permeate throughout the repo (i.e. visualizations, IoU functions, NMS, etc), which togethor sum to a significant effort for a minority user base.
> 
> In any case we aim to make YOLOv5 as broadly applicable as feasibly possible.

Would it be feasible to perform some kind of bounding box level augmentation (rotation). In other words, we could rotate the input image by X radians and also add X to the training label. This would allow us to add functionality without the need for a new dataset. 

Could this work?

**Comment by glenn-jocher** [comment]
@jmayank23 oh yeah, that's a really interesting idea! We already have many of the elements required in place for this to work:

- We have COCO available in segmentation labels in https://github.com/ultralytics/yolov5/releases/download/v1.0/coco2017labels-segments.zip
- The above segment points are augmented just like the box points in the YOLOv5 dataloader
- After augmentation the minimum rectangle is computed from the segments and training proceeds normally from there on out.

![YOLOv5 augmentation](https://user-images.githubusercontent.com/26833433/120903877-288a5a00-c649-11eb-8622-474440a084f2.jpg)

To create a fully functional rotation predictor you'd need to add a 5th value to the box regression, i.e. theta between 0 and 1, though you'd want to be careful to avoid discontinuities at the 0-360 deg break. Of course you'd also need to add the 5th term to the loss function somehow, maybe by improving the IoU function to handle rotation (which seems like it's own serious challenge).

**Comment by mohammad69h94** [comment]
> @jmayank23 oh yeah, that's a really interesting idea! We already have many of the elements required in place for this to work:
> 
> * We have COCO available in segmentation labels in https://github.com/ultralytics/yolov5/releases/download/v1.0/coco2017labels-segments.zip
> * The above segment points are augmented just like the box points in the YOLOv5 dataloader
> * After augmentation the minimum rectangle is computed from the segments and training proceeds normally from there on out.
> 
> ![YOLOv5 augmentation](https://user-images.githubusercontent.com/26833433/120903877-288a5a00-c649-11eb-8622-474440a084f2.jpg)
> 
> To create a fully functional rotation predictor you'd need to add a 5th value to the box regression, i.e. theta between 0 and 1, though you'd want to be careful to avoid discontinuities at the 0-360 deg break. Of course you'd also need to add the 5th term to the loss function somehow, maybe by improving the IoU function to handle rotation (which seems like it's own serious challenge).

How to know the detected object is rotated in images? I need to know did the image rotated or not, how to know this?

**Comment by satpalsr** [comment]
Hey @chmorgan, curious to know what did you end up using at last?

**Comment by chmorgan** [comment]
Hi @satpalsr! If I recall correctly we ended up going with a CNN and including the object orientation in our training pipeline. I think it was a good suggestion to not use yolo + an additional classifier as it would have meant more complexity, at least that's what the much more experienced AI/ML people I work with said.

**Comment by davidsvaughn** [comment]
https://github.com/hukaixuan19970627/yolov5_obb

**Comment by glenn-jocher** [comment]
@davidsvaughn thanks for sharing the repository! The Oriented Bounding Box (OBB) implementation for YOLOv5 looks really interesting. I'll make sure to check it out and see how it could be beneficial for our community. Thanks again for sharing!
